### DISCUSSION TITLE
REMOTES не все запросы выполняются и ломается счетчик ID

### DISCUSSION THREAD

**Discussion by th3pr0ust** [totumonline/totum-ru-issues-and-discussions#186] [ОШБК!] [answered]
при выполни запросов  REMOTES 
такого формата 
`data = {
    "auth": {
        "login": config.TOTUM_API_LOGIN,
        "password": config.TOTUM_API_PASSWORD
    },
    "remotes": [
        {
            "name": config.TOTUM_CREATE_ISSUE_SCRIPT_NAME_JSON,
            "data": script_data
        }
    ]
}
`

по роуту /JSON
периодически встречаем такое вот 
![Изображение вставлено в 2022-8-10 17-59](https://user-images.githubusercontent.com/110972319/183862541-008af219-28d5-4597-b56d-0575f785ce1b.jpg)

мне надо понять почему записей нет(тут конеретно мы видим в тотуме нет 5167,5163
Т.е. записи отправлены, но не получены

1) нужно понять как отловить такие ошибки
2) логируется ли это где-то
3) есть ли какое либо особое сообщение о ошибке при подобных раскладах или это тоже 200 ОК и не отслеживается ?

**Comment by ttm-alexeypolunin** [comment, marked as answer]
### Скорее всего:

Чаще всего такой эффект может быть в случае, когда по одному приходит несколько запросов, и они пишутся в одну таблицу. У таблиц по умолчанию стоит перезапуск транзакций при ошибке одновременного доступа — при запросе ремоута возможно 5 перезапусков, потом он выдаст 200 с содержимым error или в (некоторых случаях) текстом ошибки.

Тк `id` тотума резервируется при начале первого трая, то он получается пропущен.

Если ремоут завершился успешно и у вас стоит тип ответа `success/error` то при успешном завершении будет `200` + `success`.

### В случае, когда вы из внешней системы много чего пишете в одну таблицу:

— лучше на нее не вешать коды действия по добавлению, а обрабатывать асинхронно через простановку галочек кроном.
— для таблицы записи поставить параметр Актуальность https://ru.docs.totum.online/tables#table-settings-actual в значение **! Без отслеживания изменений**

### Как сделать логгирование ошибок в Ремоутах

Используйте tryCatch  https://ru.docs.totum.online/functions#fn-tryCatch в коде ремоута:

```
=catch: "remote_code_action_here"
    catch: insert(table: 'table_for_errors'; field: 'data_field_for_error' = $#exception) 
```

**Comment by th3pr0ust** [comment]
а есть вариант сделать чтобы подобные ошибки отдавались с кодом ошибки?
ну т.е. не 200
т.к. 200 это хорошо, а получается НЕ хорошо

**Reply by ttm-alexeypolunin** [reply]
Принято, добавим вариант 200/500 и сейчас еще смотрят — там какой-то модный хвостик был забетонирован, для того, что бы можно было самим написать логику кодов ответа.

**Reply by ttm-alexeypolunin** [reply]
Тут важное замечание: JSON-api всегда возвращает 200. Если вам нужно вернуть заголовки, то надо подключаться напрямую к ремоуту: `HOST/Remotes/remote_name`.

В этом случае используется `remotes_user` от которого будет выполнен ремоут, а авторизация передается либо в теле запроса, либо в заголовке. Прочитать заголовки в коде ремоута можно из переменной `$#headers`.

В этом случае работает тип ответа, заданный в поле `return` в строке ремоута в таблице ремоутов.

Если вы там поставите вариант `headers + body`, то тогда сможете сформировать нужные вам заголовки:

```
=catch: json`{"headers":$headers_ok,"body":$body_ok}`

headers_ok: $#lc
body_ok: $code_result
code_result: "code_action_here"

catch: json`{"headers":$headers_err,"body":$body_err}`

headers_err: json`[$error]`
  error: "HTTP/1.0 500 Some custom error"
body_err: $#exception
```

Таким образом мы можем отдавать стороннему сервису нужные ему коды ошибок. 

**PS. Работать будет после следующего обновления (сег-завтра)**

**Comment by th3pr0ust** [comment]
у нас запросы на добавление в БД
без каких либо расчетов 
все это идет из телеграм бота
 
можно конечно попробовать вынести задачи по добавлению в очередь, но надо надо этим подумать

**Reply by ttm-alexeypolunin** [reply]
Угу — поставьте **Без отслеживания изменений** — возможно вам еще надо увеличить количество одновременных коннектов к БД: https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/10
